### PR TITLE
BUG: Groupby.agg with reduction function with tz aware data

### LINKED ISSUE #21603: BUG: groupby.first/last loses timezone information followup
xref https://github.com/pandas-dev/pandas/pull/21573#discussion_r197425013
```
In [2]: df = pd.DataFrame({'group': [1, 1, 2],
                           'category_string': pd.Series(list('abc')).astype('category'),
                           'datetimetz': pd.date_range('20130101', periods=3, tz='US/Eastern')})
In [3]: df.groupby('group').first()
Out[3]: 
      category_string          datetimetz                                                
group                                                                                                                                                     
1                   a 2013-01-01 05:00:00                                                                      
2                   c 2013-01-03 05:00:00 
```

The example above passes data through the `first/last` compat method which strips timezone information. PR #15885 (now closed) should fix this issue (and offer a performance boost to Categorial data as mentioned in #19026)

### LINKED ISSUE #23683: A DataFrame including Timestamp with time zone fails to agg(), and makes errors.
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd

df = pd.DataFrame({
    'tag': [1,1],
    'date': [
        pd.Timestamp('2018-01-01', tz='UTC'),
        pd.Timestamp('2018-01-02', tz='UTC')]
})
df.groupby('tag').agg({'date': lambda e: e.head(1)})
```

#### Problem description

The above code makes the following errors.

```
Traceback (most recent call last):
  File "~/.pyenv/versions/anaconda3-5.3.0/lib/python3.7/site-packages/pandas/core/groupby/groupby.py", line 2670, in agg_series
    return self._aggregate_series_fast(obj, func)
  File "~/.pyenv/versions/anaconda3-5.3.0/lib/python3.7/site-packages/pandas/core/groupby/groupby.py", line 2689, in _aggregate_series_fast
    dummy)
  File "pandas/_libs/reduction.pyx", line 334, in pandas._libs.reduction.SeriesGrouper.__init__
  File "pandas/_libs/reduction.pyx", line 347, in pandas._libs.reduction.SeriesGrouper._check_dummy
ValueError: Dummy array must be same dtype

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "~/.pyenv/versions/anaconda3-5.3.0/lib/python3.7/site-packages/pandas/core/groupby/groupby.py", line 3495, in aggregate
    return self._python_agg_general(func_or_funcs, *args, **kwargs)
  File "~/.pyenv/versions/anaconda3-5.3.0/lib/python3.7/site-packages/pandas/core/groupby/groupby.py", line 1068, in _python_agg_general
    result, counts = self.grouper.agg_series(obj, f)
  File "~/.pyenv/versions/anaconda3-5.3.0/lib/python3.7/site-packages/pandas/core/groupby/groupby.py", line 2672, in agg_series
    return self._aggregate_series_pure_python(obj, func)
  File "~/.pyenv/versions/anaconda3-5.3.0/lib/python3.7/site-packages/pandas/core/groupby/groupby.py", line 2706, in _aggregate_series_pure_python
    raise ValueError('Function does not reduce')
ValueError: Function does not reduce

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "~/.pyenv/versions/anaconda3-5.3.0/lib/python3.7/site-packages/pandas/core/groupby/groupby.py", line 4656, in aggregate
    return super(DataFrameGroupBy, self).aggregate(arg, *args, **kwargs)
  File "~/.pyenv/versions/anaconda3-5.3.0/lib/python3.7/site-packages/pandas/core/groupby/groupby.py", line 4087, in aggregate
    result, how = self._aggregate(arg, _level=_level, *args, **kwargs)
  File "~/.pyenv/versions/anaconda3-5.3.0/lib/python3.7/site-packages/pandas/core/base.py", line 490, in _aggregate
    result = _agg(arg, _agg_1dim)
  File "~/.pyenv/versions/anaconda3-5.3.0/lib/python3.7/site-packages/pandas/core/base.py", line 441, in _agg
    result[fname] = func(fname, agg_how)
  File "~/.pyenv/versions/anaconda3-5.3.0/lib/python3.7/site-packages/pandas/core/base.py", line 424, in _agg_1dim
    return colg.aggregate(how, _level=(_level or 0) + 1)
  File "~/.pyenv/versions/anaconda3-5.3.0/lib/python3.7/site-packages/pandas/core/groupby/groupby.py", line 3497, in aggregate
    result = self._aggregate_named(func_or_funcs, *args, **kwargs)
  File "~/.pyenv/versions/anaconda3-5.3.0/lib/python3.7/site-packages/pandas/core/groupby/groupby.py", line 3627, in _aggregate_named
    raise Exception('Must produce aggregated value')
Exception: Must produce aggregated value
```

Actually, it works if I remove tz from the timestamps like this. So I guess it is a bug.

```
df = pd.DataFrame({
    'tag': [1,1],
    'date': [
        pd.Timestamp('2018-01-01'),
        pd.Timestamp('2018-01-02')]
})
df.groupby('tag').agg({'date': lambda e: e.head(1)})
```

#### Expected Output

```
          date
tag
1 2018-01-01 00:00:00+00:00
```

#### Output of ``pd.show_versions()``

<details>

```
>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.0.final.0
python-bits: 64
OS: Linux
OS-release: 4.15.0-34-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.4
pytest: 3.8.0
pip: 18.1
setuptools: 40.2.0
Cython: 0.28.5
numpy: 1.15.1
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 6.5.0
sphinx: 1.7.9
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: 1.2.1
tables: 3.4.4
numexpr: 2.6.8
feather: None
matplotlib: 2.2.3
openpyxl: 2.5.6
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.1.0
lxml: 4.2.5
bs4: 4.6.3
html5lib: 1.0.1
sqlalchemy: 1.2.11
pymysql: None
psycopg2: 2.7.5 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
```

</details>

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #23683
- [x] closes #21603
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -768,7 +770,17 @@ def _try_cast(self, result, obj, numeric_only=False):
...
+                    tz_dtype = None
+                    if is_datetime64tz_dtype(dtype):
+                        # GH 23683
+                        # Prior results were generated in UTC. Ensure
+                        # We localize to UTC first before converting to
+                        # the target timezone
+                        tz_dtype = dtype
+                        dtype = DatetimeTZDtype(tz='UTC')
                     result = obj._values._from_sequence(result, dtype=dtype)
+                    if tz_dtype is not None:
```
Comment (WillAyd) [inline pandas/core/groupby/groupby.py]: Looks generally good. Is this something you see us needing to do elsewhere besides just this block in groupby (no idea myself, just asking)

Comment (mroeschke) [reply pandas/core/groupby/groupby.py]: It appears that `_try_cast` is the most downstream recipient of results from the cython dispatch. Not too familiar with the groupby ops though.

But this generally needs to be down when datetimetz data goes through cython routines.

File: pandas/core/groupby/groupby.py [hunk 2]
```diff
@@ -768,7 +770,17 @@ def _try_cast(self, result, obj, numeric_only=False):
                 # The function can return something of any type, so check
                 # if the type is compatible with the calling EA.
                 try:
+                    tz_dtype = None
```
Comment (jreback) [inline pandas/core/groupby/groupby.py]: for now let's not nest this under is_extension_array_dtype as its not pretty as written; make it a separate case.

However, am thinking we we need to have a pair of (optional) routines on EA that handle this type of thing, we are tending to do this in more and more places (IIRC recently in msgpack). IOW I think we want something like

```
serialized, dtype = EA._serialize()
original = EA._from_serialize(serialized, dtype)
```
to work generically. Note when I say serialize this doesn't mean bytes on the wire (necessarily), rather going thru a transform (e.g. here we shift to UTC then transform, then need to transform back).

If you can create an issue for discussion about this.

cc @jbrockmendel @TomAugspurger @jorisvandenbossche

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/groupby.py [hunk 3]
```diff
@@ -764,7 +766,19 @@ def _try_cast(self, result, obj, numeric_only=False):
...
+                # GH 23683
+                # Prior results _may_ have been generated in UTC.
+                # Ensure we localize to UTC first before converting
+                # to the target timezone
+                try:
+                    utc_dtype = DatetimeTZDtype(tz='UTC')
+                    result = obj._values._from_sequence(result,
+                                                        dtype=utc_dtype)
+                    result = result.astype(dtype)
+                except TypeError:
```
Comment (jreback) [inline pandas/core/groupby/groupby.py]: when does this hit the type error? can you add a comment about what types of returns yield this (assume it is when the transform function is doing a non-time transform). is this hit in the tests?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/groupby.py (hunk 3 above)
Comment (mroeschke) [reply pandas/core/groupby/groupby.py]: Added a comment when this was hit (`_try_cast` is called downstream where the result is already correctly localized). Tested here:

https://github.com/pandas-dev/pandas/blob/b2c751985ebd09b72d917d08dc06193dc0922018/pandas/tests/groupby/test_transform.py#L840

**Comment by mroeschke** [comment]
All green.

File: pandas/core/groupby/groupby.py [hunk 4]
```diff
@@ -764,7 +766,21 @@ def _try_cast(self, result, obj, numeric_only=False):
...
-            if is_extension_array_dtype(dtype):
+            if is_datetime64tz_dtype(dtype):
+                # GH 23683
+                # Prior results _may_ have been generated in UTC.
+                # Ensure we localize to UTC first before converting
+                # to the target timezone
+                try:
+                    utc_dtype = DatetimeTZDtype(tz='UTC')
+                    result = obj._values._from_sequence(result,
+                                                        dtype=utc_dtype)
```
Comment (jreback) [inline pandas/core/groupby/groupby.py]: can't you just pass ``dtype='UTC'`` instead of creating the actual dtype? (as that we rarely do for timezones)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/groupby.py (hunk 4 above)
Comment (mroeschke) [reply pandas/core/groupby/groupby.py]: Just tried and passing `dtype='UTC'` didn't work here. (IMO it's more explicit too that we're passing a dtype object here)

Comment (jreback) [reply pandas/core/groupby/groupby.py]: why doesn’t this work?  that is very odd 

we rarely actually pass the dtype object anywhere else

Comment (jreback) [reply pandas/core/groupby/groupby.py]: maybe you actually need to construct as naive then localize? (these should be the same though)

Comment (mroeschke) [reply pandas/core/groupby/groupby.py]: Some tests were failing; possibly the order of operations are different in `_from_sequence`.

Using `tz_localize('UTC').tz_convert(dtype.tz)` worked though

File: pandas/core/groupby/groupby.py [hunk 5]
```diff
@@ -764,7 +766,19 @@ def _try_cast(self, result, obj, numeric_only=False):
...
-            if is_extension_array_dtype(dtype):
+            if is_datetime64tz_dtype(dtype):
+                # GH 23683
+                # Prior results _may_ have been generated in UTC.
+                # Ensure we localize to UTC first before converting
+                # to the target timezone
+                try:
+                    result = obj._values._from_sequence(result)
+                    result = result.tz_localize('UTC').tz_convert(dtype.tz)
+                except TypeError:
```
Comment (jreback) [inline pandas/core/groupby/groupby.py]: pass a dtype to DTA._from_sequence is equivalent to localizing right? I think this is true for DTI 

cc @jbrockmendel @TomAugspurger

Comment (mroeschke) [reply pandas/core/groupby/groupby.py]: It _should_, but I don't think DTA._from_sequence coerces `'UTC'` to a dtype downstream:

```
-> result = obj._values._from_sequence(result, dtype='UTC')
(Pdb) n
TypeError: data type 'UTC' not understood
```

Comment (jreback) [reply pandas/core/groupby/groupby.py]: hmm this is a bug can u create an issue

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@mroeschke lgtm. pls rebase (conflict in whatsnew) and merge on green.

**Comment by mroeschke** [comment]
Thanks @jreback

**Pull request merged**